dir.c

```diff
@@ -1795,18 +1838,28 @@ static enum path_treatment read_directory_recursive(struct dir_struct *dir,
 			subdir_state =
 				read_directory_recursive(dir, path.buf,
 							 path.len, ud,
-							 check_only, pathspec);
+							 check_only, stop_first_file, pathspec);
+
 			if (subdir_state > dir_state)
 				dir_state = subdir_state;
 		}
 
 		if (check_only) {
+
 			/* abort early if maximum state has been reached */
-			if (dir_state == path_untracked) {
-				if (cdir.fdir)
-					add_untracked(untracked, path.buf + baselen);
+			if (dir_state > max_dir_state)
+				dir_state = max_dir_state;
+
+			if (dir_state == max_dir_state) {
+				if (dir_state == path_untracked) {
+					if (cdir.fdir) {
+						add_untracked(untracked, path.buf + baselen);
+					}
```

Comment:
Don't need curly bracers for this inner-most `if`.  I believe they are even consistent about never having curly bracers for single line statements.